### PR TITLE
feat: expose live session timezone on each DataSourceConnection

### PR DESCRIPTION
**Pull request by Niels-b**
Add a `get_session_timezone()` accessor (cached, with UTC fallback) on the base `DataSourceConnection`, plus per-adapter `_fetch_session_timezone()` overrides that issue the appropriate vendor SQL: SHOW timezone (postgres/redshift), SHOW PARAMETERS (snowflake), current_timezone() (databricks/trino), current_setting (duckdb), SYSDATETIMEOFFSET() (the SQL Server family — works on SQL Server proper as well as Fabric DW and Synapse where CURRENT_TIMEZONE_ID is unsupported). BigQuery, Athena, Dremio and SparkDF return UTC directly.

This is purely additive infrastructure for the diagnostics-warehouse value-mapper layer to canonicalize TIMESTAMP_TZ values across heterogeneous source/target pairs. No existing behavior changes — nothing reads from the new method yet.

A new integration test under soda-tests/tests/integration verifies for every adapter that (a) the call returns a usable tzinfo, (b) repeat calls are cached, (c) the live session TZ matches an EXPECTED_SESSION_TIMEZONES mapping (catches silent drift), and (d) the underlying SQL produces no warning/error log records.

## Description

Please provide a description of your changes:

- What problem are you solving?
- Any expected impact on downstream packages/services?
- Reference issue # if available.

## Checklist

- [ ] I added a test to verify the new functionality.
- [ ] I verified this PR does not break [soda-extensions][1].

[1]: (https://github.com/sodadata/soda-extensions/actions/workflows/main.workflow.yaml)

### CODE REVIEW

File: soda-core/src/soda_core/common/data_source_connection.py [hunk 1]
```diff
@@ -3,17 +3,139 @@
...
 
 from tabulate import tabulate
 
+# IANA names that mean "UTC, no DST, no historical offsets". When ZoneInfo returns one
+# of these, we collapse to ``timezone.utc`` so adapter outputs are uniform regardless of
+# whether the driver reports the literal string ``"UTC"`` or one of the IANA aliases
+# below. The Etc/GMT entries are tricky — note that Etc/GMT+0 / Etc/GMT-0 are both
+# zero-offset variants (Etc/GMT+N has *negative* offset N hours per POSIX inversion,
+# but +0/-0 collapse to UTC). GMT0 is a System V-style zero-offset zone alias.
+_UTC_ALIAS_ZONE_KEYS = frozenset(
```
Comment (paulteehan) [inline soda-core/src/soda_core/common/data_source_connection.py:26]: 😮‍💨  wow this turned out to be finicky

File: soda-core/src/soda_core/common/data_source_connection.py [hunk 2]
```diff
@@ -3,17 +3,139 @@
...
+    depends on this by-identity (``is timezone.utc``) check; treat the normalization
+    as future-proofing for any fast-path branches that may want to skip ``astimezone``
+    on a guaranteed-UTC tzinfo, and as cosmetic uniformity rather than a correctness
+    fix.
+
+    Empty / None / whitespace-only input is treated as "no session TZ configured" and
+    returns ``timezone.utc`` silently. This is the conventional fallback every adapter
+    uses on a missing-row query result; promoting it to an error would emit warnings on
+    every cold start.
+
```
Comment (paulteehan) [inline soda-core/src/soda_core/common/data_source_connection.py:70]: this comment is a bit suspicious to me, I get the reasoning for not surfacing false positive warnings, but does this not open the possibility to false negatives, i.e. returning UTC when we should be throwing an error or a warning?  I don't know enough about the flow to meaningfully comment, just something to think about

File: soda-core/src/soda_core/common/data_source_connection.py [hunk 3]
```diff
@@ -29,10 +151,28 @@ def __init__(
...
+        if self._session_timezone_cache is None:
+            try:
+                self._session_timezone_cache = self._fetch_session_timezone()
+            except Exception as e:
+                logger.warning(f"Failed to query session timezone on '{self.name}'; defaulting to UTC: {e}")
+                self._session_timezone_cache = timezone.utc
+        return self._session_timezone_cache
+
+    def _fetch_session_timezone(self) -> tzinfo:
+        return timezone.utc
```
Comment (paulteehan) [inline soda-core/src/soda_core/common/data_source_connection.py]: wondering if this should raise a NotImplementedError rather than silently defaulting to UTC

File: soda-databricks/src/soda_databricks/common/data_sources/databricks_data_source.py [hunk 4]
```diff
@@ -175,9 +175,17 @@ def get_soda_data_type_name_by_data_source_data_type_names(
...
-            "timestamp": SodaDataTypeName.TIMESTAMP,
+            # Native ``timestamp`` in Databricks is TZ-aware in semantics (stored as
+            # an instant, displayed in session TZ — see Databricks "Timestamp" docs).
+            # The truly-naive variant is ``timestamp_ntz`` (Databricks Runtime 13.3+).
+            # Map ``timestamp`` to TIMESTAMP_TZ so cross-source DWH dispatch picks the
+            # TZ-aware value mapper for Databricks-as-source, otherwise driver-returned
+            # tz-aware values would be sent through the naive mapper which strips
+            # tzinfo, leaving a session-local wallclock in the warehouse instead of
+            # the canonical UTC instant.
+            "timestamp": SodaDataTypeName.TIMESTAMP_TZ,
```
Comment (paulteehan) [inline soda-databricks/src/soda_databricks/common/data_sources/databricks_data_source.py:186]: This was broken before, huh?   good thing you did this overhaul

File: soda-sqlserver/tests/unit/test_sqlserver_session_timezone.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    cursor.close.return_value = None
+
+    pyodbc_conn = MagicMock()
+    pyodbc_conn.cursor.return_value = cursor
+    instance.connection = pyodbc_conn
+    return instance, cursor
+
+
+class TestSqlServerSessionTimezoneNormalization:
+    def test_zero_offset_returns_timezone_utc_identity(self) -> None:
```
Comment (paulteehan) [inline soda-sqlserver/tests/unit/test_sqlserver_session_timezone.py:47]: the complexity here is making we think we need several CI instances of each database, each one configured with different session timezones.  obviously that's not feasible but I wonder if we could do something like that with docker containers for local testing.  It's hard to be confident we've covered all the edge cases we'd find in the wild with our relatively limited population of test database instances.  Bigger question for future

File: soda-tests/tests/integration/test_session_timezone.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    # Snapshot replay returns ``timezone.utc`` (recorded when the CI Snowflake account
+    # was on UTC); live runs against the current CI account return
+    # ``America/Los_Angeles``. Both are correct values that can appear depending on
+    # whether snapshots are active for this run, so accept either. A future flip of
+    # the CI account would still need a deliberate update here — this is not a free
+    # pass for arbitrary drift.
+    "snowflake": (ZoneInfo("America/Los_Angeles"), timezone.utc),
+    "bigquery": timezone.utc,  # BigQuery is UTC-only by design
+    "athena": timezone.utc,  # Athena is UTC-only
+    "fabric": timezone.utc,  # Inherits SQL Server adapter; test instance is UTC
```
Comment (paulteehan) [inline soda-tests/tests/integration/test_session_timezone.py:46]: I think anything that depends on the test instance state should be passed in as an environment variable.  Fine to make it default  to the instance current value, just for better maintainability

**Review by paulteehan — APPROVED** [review]
Very nice, a tricky one indeed.  Left a few comments for you to consider, nothing blocking.

File: soda-core/src/soda_core/common/data_source_connection.py [hunk 7]
```diff
@@ -3,17 +3,139 @@
...
 from abc import ABC, abstractmethod
 from collections.abc import Iterator
+from datetime import datetime, timedelta, timezone, tzinfo
 from typing import Any, Callable, Optional
 
+try:
+    from zoneinfo import ZoneInfo, ZoneInfoNotFoundError
+except ImportError:
+    ZoneInfo = None
+    ZoneInfoNotFoundError = Exception
```
Comment (mivds) [inline soda-core/src/soda_core/common/data_source_connection.py]: Why do we need `try: ... except ImportError` here? From what I find `zoneinfo` is part of the standard library and should be available since python 3.9?

File: soda-core/src/soda_core/common/data_source_connection.py [hunk 8]
```diff
@@ -3,17 +3,139 @@
...
+    A2: junk like ``'Mars/Olympus'`` previously degraded to UTC with only a debug-level
+    log; now it surfaces clearly so adapter authors and operators see the problem). The
+    single legitimate fallback site is ``DataSourceConnection.get_session_timezone()``,
+    which catches and logs one warning before returning UTC.
+    """
+    if value is None or not str(value).strip():
+        return timezone.utc
+    # Strip outer whitespace, outer quotes, then inner whitespace. Drivers may quote
+    # their result and the body of those quotes may also be empty / whitespace-only.
+    text = value.strip().strip("'\"").strip()
```
Comment (mivds) [inline soda-core/src/soda_core/common/data_source_connection.py]: Could be simplified as:
```suggestion
    text = value.strip(" \r\n\t'\"")
```

File: soda-core/src/soda_core/common/data_source_connection.py (hunk 1 above)
Comment (Niels-b) [reply soda-core/src/soda_core/common/data_source_connection.py:26]: Ooh yes, this whole operation was 😅

File: soda-core/src/soda_core/common/data_source_connection.py (hunk 2 above)
Comment (Niels-b) [reply soda-core/src/soda_core/common/data_source_connection.py:70]: We can probably make it more explicit to return a `NoTimezoneDefined` kind of thing, but that would still result in `UTC` being used further down the line. We're just moving the responsibility and complexity more upstream in that case.

It already throws an error when there is a timezone that we couldn't parse somehow (`ValueError(f"could not parse session timezone {text!r}")`).

The code is only used in the datasourceconnection in `_fetch_session_timezone`, which already handles such edge cases. The `parse_session_timezone` is a "helper" function

File: soda-core/src/soda_core/common/data_source_connection.py (hunk 3 above)
Comment (Niels-b) [reply soda-core/src/soda_core/common/data_source_connection.py]: Hmm yes, I think that's a good idea. Especially when we add new datasources 👍

File: soda-sqlserver/tests/unit/test_sqlserver_session_timezone.py (hunk 5 above)
Comment (Niels-b) [reply soda-sqlserver/tests/unit/test_sqlserver_session_timezone.py:47]: For some datasources, yes this definitely should be possible with containers. But I think that's out of scope for this PR to set up

File: soda-tests/tests/integration/test_session_timezone.py (hunk 6 above)
Comment (Niels-b) [reply soda-tests/tests/integration/test_session_timezone.py:46]: Not really a fan of splitting it up to env variables (we'd also have to do some more conversion since we're passing a `timezone.utc` now). I want it to fail "hard" and that we have to fix it in the code itself, rather than in a separate location in the env variables (thinking we may forget to do/update this).

File: soda-core/src/soda_core/common/data_source_connection.py [hunk 9]
```diff
@@ -3,17 +3,139 @@
...
+    fix.
+
+    Empty / None / whitespace-only input is treated as "no session TZ configured" and
+    returns ``timezone.utc`` silently. This is the conventional fallback every adapter
+    uses on a missing-row query result; promoting it to an error would emit warnings on
+    every cold start.
+
+    Raises ``ValueError`` on non-empty unparseable input (the colleague-review concern
+    A2: junk like ``'Mars/Olympus'`` previously degraded to UTC with only a debug-level
+    log; now it surfaces clearly so adapter authors and operators see the problem). The
```
Comment (mivds) [inline soda-core/src/soda_core/common/data_source_connection.py]: The stuff in parentheses feels like noise that doesn't belong in docstring

File: soda-core/src/soda_core/common/data_source_connection.py [hunk 10]
```diff
@@ -3,17 +3,139 @@
...
+    single legitimate fallback site is ``DataSourceConnection.get_session_timezone()``,
+    which catches and logs one warning before returning UTC.
+    """
+    if value is None or not str(value).strip():
```
Comment (mivds) [inline soda-core/src/soda_core/common/data_source_connection.py]: The `str(value).strip()` clause here seems to strictly be a duplicate of the clause below, no?

File: soda-core/src/soda_core/common/data_source_connection.py [hunk 11]
```diff
@@ -3,17 +3,139 @@
...
+        return timezone.utc
+    # Strip outer whitespace, outer quotes, then inner whitespace. Drivers may quote
+    # their result and the body of those quotes may also be empty / whitespace-only.
+    text = value.strip().strip("'\"").strip()
+    if not text:
+        # Outer quotes wrapped an empty (or whitespace-only) string — same intent as
+        # a raw-empty input on the conventional empty-fallback path.
+        return timezone.utc
+    if text.upper() in ("UTC", "GMT", "Z"):
+        return timezone.utc
```
Comment (mivds) [inline soda-core/src/soda_core/common/data_source_connection.py:87]: Why is this a separate case from the `ZoneInfo` parsing below?

File: soda-core/src/soda_core/common/data_source_connection.py [hunk 12]
```diff
@@ -3,17 +3,139 @@
...
+
+    if text.lower() in _HOST_LOCAL_LITERAL_VALUES:
+        # Driver said "use the host local TZ". Resolve via the OS now. The result is a
+        # fixed-offset tzinfo derived from the current local time — DST drift across a
+        # long-lived connection is the same as for SQL Server / DB2 (out of scope per
+        # the team's "DWH connections are short-lived" stance), but the wallclock at
+        # canonicalization time is correct.
+        return datetime.now().astimezone().tzinfo
+
+    if ZoneInfo is not None:
```
Comment (mivds) [inline soda-core/src/soda_core/common/data_source_connection.py]: Can be dropped if the import structure above is changed

File: soda-core/src/soda_core/common/data_source_connection.py [hunk 13]
```diff
@@ -3,17 +3,139 @@
...
+        try:
+            zi = ZoneInfo(text)
+        except (ZoneInfoNotFoundError, ValueError, OSError):
+            zi = None
+        if zi is not None:
+            # Collapse UTC-equivalent IANA aliases to ``timezone.utc`` so an adapter
+            # reporting ``"Etc/UTC"`` returns the same singleton as one reporting the
+            # literal ``"UTC"`` or the ``+00:00`` numeric offset. Without this, the
+            # docstring's identity-parity claim would be false for one specific input.
+            if zi.key in _UTC_ALIAS_ZONE_KEYS:
```
Comment (mivds) [inline soda-core/src/soda_core/common/data_source_connection.py]: Could possibly be replaced by looking at the offset relative to UTC rather than keeping a hardcoded list of UTC-like timezones. Not sure it's better, just something to consider.


```suggestion
            if zi.utcoffset(datetime.now()).total_seconds() == 0:
```

File: soda-databricks/src/soda_databricks/common/data_sources/databricks_data_source_connection.py [hunk 14]
```diff
@@ -28,6 +32,12 @@ def _create_connection(
             **config.to_connection_kwargs(),
         )
 
+    def _fetch_session_timezone(self) -> tzinfo:
+        with self.connection.cursor() as cursor:
+            cursor.execute("SELECT current_timezone()")
+            row = cursor.fetchone()
+        return parse_session_timezone(row[0] if row else "")
```
Comment (mivds) [inline soda-databricks/src/soda_databricks/common/data_sources/databricks_data_source_connection.py]: Nit, feel free to ignore: outcome will be the same, but we could be more explicit here:


```suggestion
        if not row:
            return datetime.timezone.utc
        return parse_session_timezone(row[0])
```

File: soda-snowflake/src/soda_snowflake/common/data_sources/snowflake_data_source_connection.py [hunk 15]
```diff
@@ -145,3 +149,33 @@ def _create_connection(
             application="Soda",
             **config.to_connection_kwargs(),
         )
+
+    def _fetch_session_timezone(self) -> tzinfo:
+        cursor = self.connection.cursor()
+        try:
```
Comment (mivds) [inline soda-snowflake/src/soda_snowflake/common/data_sources/snowflake_data_source_connection.py]: Any reason why we're not using `with self.connection.cursor()` here?

File: soda-sqlserver/src/soda_sqlserver/common/data_sources/sqlserver_data_source_connection.py [hunk 16]
```diff
@@ -196,5 +196,31 @@ def _create_connection(
...
+    def _fetch_session_timezone(self) -> tzinfo:
+        # Use SYSDATETIMEOFFSET() instead of CURRENT_TIMEZONE_ID() so the same query works
+        # across the whole SQL Server family: SQL Server proper, Microsoft Fabric DW, Azure
+        # Synapse Dedicated/Serverless. CURRENT_TIMEZONE_ID is unsupported in Fabric and
+        # Synapse Dedicated; SYSDATETIMEOFFSET is universally supported.
+        # The connection registers ``handle_datetimeoffset`` as a pyodbc output converter for
+        # SQL_TYPE -155, so the returned value is a tz-aware ``datetime`` whose tzinfo is the
+        # exact offset reported by the server.
+        cursor = self.connection.cursor()
+        try:
```
Comment (mivds) [inline soda-sqlserver/src/soda_sqlserver/common/data_sources/sqlserver_data_source_connection.py]: Same question here around using `with cursor`

**Review by mivds — APPROVED** [review]
LGTM, nice work finally tackling our timezone issues at the root.

I left some comments for consideration, but there's nothing blocking there.

File: soda-core/src/soda_core/common/data_source_connection.py (hunk 7 above)
Comment (Niels-b) [reply soda-core/src/soda_core/common/data_source_connection.py]: Yes, good catch! Also makes the comments below a bit simpler. Thank you.

File: soda-core/src/soda_core/common/data_source_connection.py (hunk 11 above)
Comment (Niels-b) [reply soda-core/src/soda_core/common/data_source_connection.py:87]: Good catch! Only the `"Z"` one should be as an exception here, the rest should be handled normally by the timezones.

File: soda-core/src/soda_core/common/data_source_connection.py (hunk 13 above)
Comment (Niels-b) [reply soda-core/src/soda_core/common/data_source_connection.py]: As discussed on Slack, this one we probably shouldn't do as it introduces a bug.

File: soda-snowflake/src/soda_snowflake/common/data_sources/snowflake_data_source_connection.py (hunk 15 above)
Comment (Niels-b) [reply soda-snowflake/src/soda_snowflake/common/data_sources/snowflake_data_source_connection.py]: with is always better; changing it. Thanks!
